sw/device/silicon_creator/rom_ext/rom_ext.c

```diff
@@ -394,143 +577,141 @@ static rom_error_t rom_ext_attestation_silicon(void) {
+      /*attest_binding=*/attest_binding, rom_ext_manifest->max_key_version));
   HARDENED_RETURN_IF_ERROR(otbn_boot_cert_ecc_p256_keygen(
-      kDiceKeyCdi0, &cdi_0_pubkey_id, &curr_attestation_pubkey));
-  hardened_bool_t cert_valid = kHardenedBoolFalse;
-  HARDENED_RETURN_IF_ERROR(cert_x509_asn1_check_serial_number(
-      dice_cert_obj.cert_body_p, /*offset=*/0,
-      (uint8_t *)cdi_0_pubkey_id.digest, &cert_valid, /*out_cert_size=*/NULL));
-  if (launder32(cert_valid) == kHardenedBoolFalse) {
-    HARDENED_CHECK_EQ(cert_valid, kHardenedBoolFalse);
+      kDiceKeyCdi0, &dice_chain.subject_pubkey_id, &dice_chain.subject_pubkey));
+
+  // Check if the current CDI_0 cert is valid.
+  HARDENED_RETURN_IF_ERROR(dice_chain_load_cert_obj("CDI_0", 6));
+  if (launder32(dice_chain.cert_valid) == kHardenedBoolFalse) {
+    HARDENED_CHECK_EQ(dice_chain.cert_valid, kHardenedBoolFalse);
     dbg_printf("CDI_0 certificate not valid. Updating it ...\r\n");
-    uint32_t updated_cert_size = kCdi0MaxCertSizeBytes;
-    HARDENED_RETURN_IF_ERROR(dice_cdi_0_cert_build(
-        (hmac_digest_t *)boot_measurements.rom_ext.data,
-        rom_ext_manifest->security_version, &cdi_0_key_ids,
-        &curr_attestation_pubkey, cdi_0_cert, &updated_cert_size));
     // Update the cert page buffer.
-    size_t cert_page_left = dice_certs_buffer_space_remaining();
-    HARDENED_RETURN_IF_ERROR(perso_tlv_cert_obj_build(
-        "CDI_0", /*needs_endorsement=*/false, cdi_0_cert, updated_cert_size,
-        dice_cert_obj.obj_p, &cert_page_left));
-    dice_certs_page_dirty = kHardenedBoolTrue;
-    // Reload the cert perso LTV object.
-    HARDENED_RETURN_IF_ERROR(perso_tlv_get_cert_obj(
-        dice_cert_obj.obj_p, dice_certs_buffer_space_remaining(),
-        &dice_cert_obj));
-    HARDENED_RETURN_IF_ERROR(rom_ext_attestation_increment_cert_offset());
-    // If the CDI_0 cert is updated, it could overrun the CDI_1 cert, so we make
-    // sure to update that as well by erasing the existing cert from the buffer.
-    memset(dice_cert_obj.obj_p, 0, dice_certs_buffer_space_remaining());
+    size_t updated_cert_size = kScratchCertSizeBytes;
+    HARDENED_RETURN_IF_ERROR(
+        dice_cdi_0_cert_build((hmac_digest_t *)attest_binding->data,
+                              rom_ext_manifest->security_version,
+                              &dice_chain.key_ids, &dice_chain.subject_pubkey,
+                              dice_chain.scratch_cert, &updated_cert_size));
+    HARDENED_RETURN_IF_ERROR(dice_chain_push_cert(
+        "CDI_0", dice_chain.scratch_cert, updated_cert_size));
   } else {
-    HARDENED_RETURN_IF_ERROR(rom_ext_attestation_increment_cert_offset());
+    // Cert is valid, move to the next one.
+    HARDENED_CHECK_EQ(dice_chain.cert_valid, kHardenedBoolTrue);
+    dice_chain_next_cert_obj();
   }
+
+  // Replace UDS with CDI_0 key for endorsing next stage cert.
+  HARDENED_RETURN_IF_ERROR(otbn_boot_attestation_key_save(
```

Comment:
we only want to do this if the cert was valid, as there is already a call to `otbn_boot_attestation_key_save()` in `dice_cdi_0_cert_build()`

Reply:
same for CDI_1 below